docs/dev/proposal-limit-all.md

```diff
@@ -0,0 +1,83 @@
+With changes specified in this document, SQL plugin clients will always be aware if SQL plugin returned all documents that match a given query and when it does not, how to retrieve all documents.
+
+Both SQL plugin and OpenSearch limit how many results are returned when paging is not used. However, SQL plugin does not communicate to the client when these limits are hit. This leaves clients with the incorrect impression that all matching data was returned.
+
+This document describes syntax changes for SQL. The same behaviour can be added to PPL but the syntax will be different.
+
+# Use Cases
+
+The following use cases should notify users when incomplete data is returned.
+* opensearchsql-cli
+* workbench
+* observability plugin
+* JDBC driver
+* ODBC driver
+
+# Definitions
+* `totalSize` &mdash; number of documents that match given query.
+* `index.max_result_window` &mdash; maximum number of documents that OpenSearch will return without a scroll request. A cluster setting.
+* `query.size_limit` &mdash; maximum number of rows that SQL plugin will return without a cursor request. A SQL plugin setting.
+
+# Assumptions
+1. Query is not an aggregation query.
+2. `query.size_limit < index.max_result_window`
+
+# Desired behaviour
+
+## Without cursor request, without LIMIT clause
+
+```sql
+SELECT <...> FROM <...> WHERE <...> 
+```
+
+## Without cursor request, with LIMIT clause
+
+```sql
+SELECT <...> FROM <...> WHERE <...> LIMIT X 
+```
+Returns up to `X` number of rows.
+
+If `X` is less than `index.max_window_size` all rows are returned.
+
+If `X` is greater than `index.max_window_size`, only first `X` rows are returned and response indicates that more data is [available](#more-data-available-response).
+
+```sql
+SELECT <...> FROM <...> WHERE <...> LIMIT X OFFSET Y
```

Comment:
Do we need to support this for legacy engine?  
If not, we can say this is not for legacy (phase II).  

... in which case, we should discuss phase I (legacy support) vs phase II (enhanced support) vs phase III (future)